drivers/usbc/tcpc/ucpd_stm32gx.c

```diff
@@ -0,0 +1,1442 @@
+	 *   MsgID      b11:9  -> running byte counter (extracted from rx msg)
+	 *   Power Role b8     -> stored in static, from set_msg_header()
+	 *   Spec Rev   b7:b6  -> PD spec revision (extracted from rx msg)
+	 *   Data Role  b5     -> stored in static, from set_msg_header
+	 *   Msg Type   b4:b0  -> data or ctrl type = PD_CTRL_GOOD_CRC
+	 */
+	/* construct header message */
+	tx_header.message_type = PD_CTRL_GOOD_CRC;
+	if (tx_type == PD_PACKET_SOP) {
+		tx_header.port_power_role = data->msg_header.pr;
+		tx_header.port_data_role = data->msg_header.dr;
+	} else {
+		tx_header.port_power_role = 0;
+		tx_header.port_data_role = 0;
+	}
+	tx_header.message_id = rx_header.message_id;
+	tx_header.number_of_data_objects = 0;
+	tx_header.specification_revision = rx_header.specification_revision;
+	tx_header.extended = 0;
+
+	/* Good CRC is header with no other objects */
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].msg_len = 2;
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].data.header =
+		tx_header.raw_value;
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].type = tx_type;
+
+	/* Notify ucpd task that a GoodCRC message tx request is pending */
+	info->evt |= UCPD_EVT_GOOD_CRC_REQ;
+}
+
+/**
+ * @brief Transmit a power delivery message
+ *
+ * @return 0 on success
+ * @return -EFAULT on failure
+ */
+static int ucpd_transmit_data(const struct device *dev,
+			      struct pd_msg *msg)
+{
+	struct tcpc_data *data = dev->data;
+
+	/* Length in bytes = (4 * object len) + 2 header bytes */
+	int len = PD_CONVERT_PD_HEADER_COUNT_TO_BYTES(msg->header.number_of_data_objects) + 2;
+
+	if (len > UCPD_BUF_LEN) {
+		return -EFAULT;
+	}
+
+	/* Store tx msg info in TCPM msg descriptor */
+	data->ucpd_tx_buffers[TX_MSG_TCPM].msg_len = len;
+	data->ucpd_tx_buffers[TX_MSG_TCPM].type = msg->type;
+	data->ucpd_tx_buffers[TX_MSG_TCPM].data.header = msg->header.raw_value;
+
+	/* Copy msg objects to ucpd data buffer, after 2 header bytes */
+	memcpy(data->ucpd_tx_buffers[TX_MSG_TCPM].data.msg + 2,
+	       (uint8_t *)msg->data, len - 2);
+
+	/*
+	 * Check for hard reset message here. A different event is used for hard
+	 * resets as they are able to interrupt ongoing transmit, and should
+	 * have priority over any pending message.
+	 */
+	if (msg->type == PD_PACKET_TX_HARD_RESET) {
+		data->alert_info.evt = UCPD_EVT_HR_REQ;
+	} else {
+		data->alert_info.evt = UCPD_EVT_TCPM_MSG_REQ;
+	}
+
+	/* Start transmission */
+	k_work_submit(&data->alert_info.work);
+
+	return 0;
+}
+
+/**
+ * @brief Tests if a received Power Delivery message is pending
+ *
+ * @return true if message is pending, else false
+ * @return -EIO on failure
+ */
+static bool ucpd_is_rx_pending_msg(const struct device *dev,
+				   enum pd_packet_type *type)
+{
+	struct tcpc_data *data = dev->data;
+	int ret = 0;
+
+	ret = (*(uint32_t *)data->ucpd_rx_buffer > 0);
+
+	if (ret & (type != NULL)) {
+		*type = *(uint16_t *)data->ucpd_rx_buffer;
+	}
+
+	return ret;
```

Comment:
function returls bool, but `int` is used here

Reply:
Done